compiler/src/iree/compiler/Dialect/Stream/Transforms/FoldUniformOperands.cpp

```diff
@@ -248,8 +268,118 @@ inlineUniformConstants(mlir::FunctionOpInterface funcOp,
-  entryBlock.eraseArguments(
-      [&](BlockArgument arg) { return deadArgMap.test(arg.getArgNumber()); });
+  entryBlock.eraseArguments(deadArgMap);
+}
+
+// Deduplicates workloads that have the same value at all dispatch sites.
+//
+// Example:
+//   stream.cmd.dispatch @foo[%0, %0](...)
+//  ->
+//   stream.cmd.dispatch @foo[%0](...)
+// + deduped arguments in `stream.executable.export`
+// + deduped ordinals in `dispatch.workload.ordinal`
+static void
+deduplicateWorkloads(IREE::Stream::ExecutableExportOp exportOp,
+                     SmallVector<IREE::Stream::CmdDispatchOp> &dispatchOps) {
+  if (exportOp.getWorkgroupCount().empty()) {
+    return;
+  }
+  mlir::FunctionOpInterface funcOp = exportOp.lookupFunctionRef();
+  if (!funcOp) {
+    return;
+  }
+  IREE::Stream::CmdDispatchOp anyDispatchOp = dispatchOps.front();
+  unsigned workloadCount = anyDispatchOp.getWorkload().size();
+
+  // Build a map of workload indices to its base duplicate for each dispatch
+  // site. Base/non-duplicated values will be identity.
+  // Example: (%a, %b, %a, %b) -> (0, 1, 0, 1)
+  static const int kUnassigned = -1;
+  SmallVector<SmallVector<unsigned>> dupeIndexMaps(dispatchOps.size());
+  for (auto [i, dispatchOp] : llvm::enumerate(dispatchOps)) {
+    SmallVector<unsigned> &dupeIndexMap = dupeIndexMaps[i];
+    dupeIndexMap.resize(workloadCount, kUnassigned);
+    SmallVector<Value> workloads = dispatchOp.getWorkload();
+    for (unsigned i = 0; i < workloads.size(); ++i) {
+      for (unsigned j = 0; j < i; ++j) {
+        if (workloads[j] == workloads[i]) {
+          dupeIndexMap[i] = j;
+          break;
+        }
+      }
+    }
+  }
+
+  // Per-workload now find which are consistently duplicated.
+  auto [uniformDupeIndexMap, deadOperandsMap] =
+      getUniformDupeIndexingAndDeadValues(dupeIndexMaps);
+  if (deadOperandsMap.none()) {
+    // No-op.
+    return;
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "deduplicateWorkloads for " << exportOp.getName() << "\n";
+    llvm::dbgs() << "  dead workloads: ";
+    llvm::interleaveComma(deadOperandsMap.set_bits(), llvm::dbgs());
+    llvm::dbgs() << "\n";
+    for (auto [deadIdx, liveIdx] : llvm::enumerate(uniformDupeIndexMap)) {
+      if (deadIdx == liveIdx) {
+        continue;
+      }
+      llvm::dbgs() << "  %arg" << deadIdx << " -> %arg" << liveIdx << "\n";
+    }
+  });
+
+  // Update the ordinals in the workload ordinal ops.
+  unsigned ordinalCount = 0;
+  SmallVector<unsigned> newOrdinals(uniformDupeIndexMap.size());
+  for (auto [deadIdx, liveIdx] : llvm::enumerate(uniformDupeIndexMap)) {
+    if (deadIdx > liveIdx) {
+      continue;
+    }
+    newOrdinals[deadIdx] = ordinalCount++;
+  }
+  funcOp->walk([&](IREE::TensorExt::DispatchWorkloadOrdinalOp ordinalOp) {
+    uint64_t ordinal = ordinalOp.getOrdinal().getZExtValue();
+    assert(ordinal < newOrdinals.size() &&
+           "the ordinal shouldn't be larger than the number of workloads");
```

Comment:
I'd drop the assertion. We already have indexing assertion in vector's implementation.